client/service/src/task_manager/mod.rs

```diff
@@ -193,12 +210,12 @@ impl SpawnEssentialTaskHandle {
 		task: impl Future<Output = ()> + Send + 'static,
 		task_type: TaskType,
 	) {
-		let mut essential_failed = self.essential_failed_tx.clone();
+		let essential_failed = self.essential_failed_tx.clone();
 		let essential_task = std::panic::AssertUnwindSafe(task)
 			.catch_unwind()
 			.map(move |_| {
 				log::error!("Essential task `{}` failed. Shutting down service.", name);
-				let _ = essential_failed.send(());
+				let _ = essential_failed.close_channel();
```

Comment:
Sending didn't work for some reason (it doesn't work in the current version of substrate!). I think it is fine as it as it is an internal detail. I also added a test to make sure that if we ever make this feature more sophisticated it will at least ensure that `fn future()` stops properly.